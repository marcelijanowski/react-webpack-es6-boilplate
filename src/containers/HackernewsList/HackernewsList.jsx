import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';

import { actions } from '../../reducers/hackernews/stories/action';
import HackernewsStory from '../../components/organisms/hackernews-story/HackernewsStory';
export class HackernewsList extends Component {
  state = {
    isLoading: false,
    items: null,
    error: null
  }
  static getDerivedStateFromProps(props, state) {
    const { hackernewsStory } = props;
    const { error, isFetching, stories } = hackernewsStory;

    if (stories) {
      return {
        items: stories,
        isLoading: false
      }
    }

    if (error) {
      return {
        error,
        isLoading: false
      }
    }
    
    return {
      isLoading: true
    }
  }
  componentDidMount() {
    const { getHackernewsStoriesRequest, type } = this.props;
    getHackernewsStoriesRequest(type);
  }

  render() {
    const { isLoading, items, error } = this.state;
    return (
      <React.Fragment>
      { error &&  (<div className="hackernewslist-state__loading">There are some problems with loading stories...</div>)}
      { isLoading  &&  (<div className="hackernewslist-state_error">Data is loading</div>) } 
      { items &&  (
        <ul class="mdc-list hackernewslist_list">
          { items.length && items.map(item => <HackernewsStory item={item} key={item.id} />)}
        </ul>
      )}
      </React.Fragment>
    )
  }
}

HackernewsList.defaultProps = {
  type: 'topStory'
};

HackernewsList.propTypes = {
  type: PropTypes.string
};

function mapStateToProps ( state ) {
  const { hackernewsStory, routing } = state;

  return {
    hackernewsStory,
    routing
  }
}

function mapDispatchToProps (dispatch) {
  return {
    getHackernewsStoriesRequest: (type) => dispatch(actions.getHackernewsStoriesRequest(type))
  }
}

export default connect(mapStateToProps, mapDispatchToProps)(HackernewsList)