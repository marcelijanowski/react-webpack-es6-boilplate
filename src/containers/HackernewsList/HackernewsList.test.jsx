import React, { PureComponent } from 'react'
import renderer from 'react-test-renderer';

import { HackernewsList } from './HackernewsList';
import { mockRef, mountComponent } from '../../../test/testHelper';

describe('HackernewsList container', () => {
  let getHackernewsStoriesRequest;
  beforeEach(() => {
    getHackernewsStoriesRequest = jasmine.createSpy('getHackernewsStoriesRequest');
  });
  it('renders a snapshot', () => {
    const tree = renderer.create(<HackernewsList 
      type="topStory"
      getHackernewsStoriesRequest={getHackernewsStoriesRequest}
      hackernewsStory={{
        error: null,
        isFetching: false,
        stories: [{
          id: 1
        }]
      }}
    />).toJSON();
    expect(tree).toMatchSnapshot();
  });
})