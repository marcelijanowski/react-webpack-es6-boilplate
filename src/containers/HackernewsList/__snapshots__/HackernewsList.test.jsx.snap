// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HackernewsList container renders a snapshot 1`] = `
<ul
  class="mdc-list hackernewslist_list"
>
  <li
    class="mdc-list-item"
  >
    <span
      class="mdc-list-item__text"
    >
      1
      <span
        class="mdc-list-item__secondary-text"
      >
        Second-line text
      </span>
    </span>
  </li>
</ul>
`;
